### ISSUE TITLE
notice: improve accessibility of notice title

### ISSUE THREAD

**Issue by JohnXLivingston** [McShelby/hugo-theme-relearn#897]
Hello,

I had an accessibility audit done on my documentation website.
They found an issue with the notice boxes: "the role of the ‘warning’, ‘tip’ and ‘info’ boxes is not communicated trough another mechanism other than visual. Because these boxes do convey information, they should have a title."

https://www.w3.org/WAI/WCAG21/Understanding/name-role-value
https://www.w3.org/WAI/WCAG21/Understanding/non-text-content.html

**Comment by McShelby** [comment]
Thanks for the info. I'ld love to merge a PR for this.

While I usually fix obvious a11y issues when I see them, I am not a expert on this.

Fixing a11y issues are costing a lot of time, scanning thru highly academic specifications I simply don't have the time for.

That's also the reason why a11y is not an {advertised](https://mcshelby.github.io/hugo-theme-relearn/) feature of this theme.

**Comment by McShelby** [comment]
I've read thru various documents and your links plus also asking AI. It is still unclear to me, what needs to be changed to pass a11y criteria.

As you can see by upgrading from Learn to Relearn, I already spent some decent amount of time to fix most obvious flaws. But here, I am clueless.

**Comment by JohnXLivingston** [comment]
> As you can see by upgrading from Learn to Relearn, I already spent some decent amount of time to fix most obvious flaws. But here, I am clueless.

Yes I saw that. I migrated from hugo-theme-learn to hugo-theme-relearn yesterday, and it fixes a lot of accessibility issues. Thanks a lot for that!
I assumed you will better know than me how to fix this issue. Sorry for that. I will ask more precision to the person who made the audit, and i will make a PR when i have an answer.

**Comment by McShelby** [comment]
That would be great. Currently I assume we just need to add some few ARIA attributes to the notice DOM. So couldn't be to hard to implement.

**Comment by JohnXLivingston** [comment]
> That would be great. Currently I assume we just need to add some few ARIA attributes to the notice DOM. So couldn't be to hard to implement.

Yes. I'm discussing with an expert. First, we considered adding a `<section>` and an `<h2>`, but this is not the good solution, because it would break navigation.

I have maybe another solution, just by adding `aria-hidden="true"` on the infobox title, and `aria-label="Tips"` on the `<p>` inside the infobox.
I'm waiting the expert advice, and if it is ok i will make the PR.

**Comment by McShelby** [comment]
[MDN](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Attributes/aria-label) quotes

> The aria-label attribute defines a string value that labels an interactive element

What bugs me is the term _interactive_ as the notice box aren't. So I was a bit suprised by the limiting scope of this attribute but couldn't find anything better.

Also, `aria-heading` seems not good, as it has the same semantic as a HTML heading.

HTML headings and therefore `aria-heading` is also not working as it will break the document outline.

That's the conclusion from my frustrating a11y exploration yesterday. And that's also the reason why its no official feature ;-)

**Comment by JohnXLivingston** [comment]
The expert says that aria-label on `p` will not be read by screen readers. Still working on a solution.

**Comment by JohnXLivingston** [comment]
So, after a long discussion, we have a partial solution...
Something like this would work:

```html
<div class="box notices cstyle>
  <div class="box-label" aria-hidden="true">Tips</div>
  <div class="box-content">
    <p>
      <span class="sr-only">Tips:</span>
      The content
    </p>
  </div>
</div>
```

The basic idea: hide the visible title for screen readers, and add an invisible text for screen readers in the paragraph.
The `sr-only` class comes with font-awesome, and is a trick to make invisible text for screen reader only. This is documented here (in french, sorry): https://www.lalutineduweb.fr/alternatives-textuelles-texte-masque-css/

but.... this works not in all cases...
If there is more than one paragraph (or any other html tag that is not "phrasing content"), this would not work.

**Comment by JohnXLivingston** [comment]
I'm not a Hugo dev, i don't know if this would be possible to do?
(even if the `p` comes from the content, which is the case in my project)

**Comment by McShelby** [comment]
Good thing, the theme already has such a CSS class `a11y-only`.

Everything inside the `box-content` is generated by Hugo. So while we could mess around with it, I would strongly avoid it.

Would moving up the a11y span also be feasible?

````
<div class="box notices cstyle>
  <div class="box-label" aria-hidden="true">Tips</div>
  <span class="sr-only">Tips:</span>
  <div class="box-content">
    <p>
      The content
    </p>
  </div>
</div>
````

or 


````
<div class="box notices cstyle>
  <div class="box-label">
    <span aria-hidden="true">Tips</span>
    <span class="sr-only">Tips:</span>
  </div>
  <div class="box-content">
    <p>
      The content
    </p>
  </div>
</div>
````

**Comment by JohnXLivingston** [comment]
No. If the title is outside the paragraph, something should say it is the title for the box-content.

The ideal solution would be to insert a heading... But the level should be the current heading level + 1. I doubt you can get this information. And it could break headings that are after the notice...
And, moreover, it would break the heading level of text that is after the notice.

Damn.
Can't believe there is no aria role for such contextual notices... That would be so much easier!

I'm out of ideas. Sorry.

**Comment by McShelby** [comment]
> No. If the title is outside the paragraph, something should say it is the title for the box-content.

The title relates to the whole box-content not just a/the first paragraph inside that box. As you already assumed, there could by multiple paragraph and also other content inside of it.

Headings is a major misconception in HTML but we have to live with it. As you already assumed, messing with the heading level will not work. It's inaccessible to the code. Also the question would arise, what to do, if the existing content is already on `h6`...

> Damn.
> I'm out of ideas. Sorry.

Now, you starting to feel just like me ;-)

Would this do the trick?

````
<div class="box notices cstyle>
  <div class="box-label" aria-hidden="true">Tips</div>
  <div class="box-content">
    <span class="sr-only">Tips:</span>
    <p>
      First paragraph of content
    </p>
    <p>
      Second paragraph of content
    </p>
  </div>
</div>
````

**Comment by JohnXLivingston** [comment]
> Now, you starting to feel just like me ;-)

To be honest, i already feeled like you yesterday. I really hoped you will have answers ;)



> Would this do the trick?

I will ask.

**Comment by McShelby** [comment]
![th-198712756](https://github.com/user-attachments/assets/c24000c6-22db-44d6-b490-e0a00aa7de11)

**Comment by JohnXLivingston** [comment]
> Anyways, thanks for making the initial issue. There's always something to learn.

Yes!
And maybe someone will read it, and propose a good solution. Who knows?

**Comment by McShelby** [comment]
@JohnXLivingston Hi John, if you are still around and have access to your a11y expert, would this be a viable solution:

````
<details class="box notices cstyle">
  <summary class="box-label">Tips</summary>
  <div class="box-content">
    <p>
      First paragraph of content
    </p>
    <p>
      Second paragraph of content
    </p>
  </div>
</details>
````

As Hugo goes into a similar direction for other reasons, I am thinking of changing the implementation regardless what the a11y expert may say. Just wanted to make sure if this solution could fix your issue.

**Comment by McShelby** [comment]
@JohnXLivingston I've pushed the changes to main, so the official [docs](https://mcshelby.github.io/hugo-theme-relearn/shortcodes/notice/index.html) already contain the new (and hopefully improved) implementation

**Comment by JohnXLivingston** [comment]
Hi @McShelby 
I will ask to the expert.

Thanks a lot!

**Comment by McShelby** [comment]
@JohnXLivingston I will close this as completed for now. If the feedback will be negative, I reopen this issue or - most likely - will create a follow-up issue.

**Comment by JohnXLivingston** [comment]
> @JohnXLivingston I will close this as completed for now. If the feedback will be negative, I reopen this issue or - most likely - will create a follow-up issue.

The expert just respond that she has no time right now.
So yes, you can close. I will try to find someone else to check the patch.

Thanks a lot for the help.
